src/arch/amdgpu/vega/insts/instructions.hh

```diff
@@ -43915,78 +43916,232 @@ namespace VegaISA
-        {
-            return numDstRegOperands() + numSrcRegOperands();
-        } // getNumOperands
 
-        int numDstRegOperands() override { return 1; }
-        int numSrcRegOperands() override { return 3; }
+        // These values and meanings are described in the MI300 ISA manual:
+        //
+        // https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/
+        //    instruction-set-architectures/
+        //    amd-instinct-mi300-cdna3-instruction-set-architecture.pdf
+        //
+        // in section 7.1.4.2. In theory, only the M, N, K, and H values change
+        // for each MFMA instruction and therefore this could be templated.
 
-        int
-        getOperandSize(int opIdx) override
-        {
-            switch (opIdx) {
-              case 0: // src0 "A"
-                return 8;
-              case 1: // src1 "B"
-                return 8;
-              case 2: // src2 "C"
-                return 32;
-              case 3: // dst
-                return 32;
-              default:
-                fatal("op idx %i out of bounds\n", opIdx);
-                return -1;
+        // Output layout
+        constexpr int H = _delta == 2 ? 1 : 4;
+        constexpr int B_I = std::ceil(64.0f / (N * M / H));
+        constexpr int M_I = (64 / B_I) / N;
+        constexpr int G = M / (H * M_I);
+
+        float result[M][N];
+
+        // Input layout
+        constexpr int K_L = K / (64 / (M * B));
+
+        for (int block = 0; block < B; block++) {
+            // Load src2 into result. src2 is row major
+            for (int i = 0; i < M; ++i) {
+                for (int j = 0; j < N; ++j) {
+                    int item = (i % H) + H * (i/(H*M_I) + G * (block / B_I));
+                    int lane = j + N * ((i / H) % M_I + M_I * (block % B_I));
+
+                    result[i][j] = src2[item][lane];
+                }
             }
-        } // getOperandSize
 
-        void execute(GPUDynInstPtr) override;
+            // Compute new result
+            for (int i = 0; i < M; ++i) {
+                for (int j = 0; j < N; ++j) {
+                    for (int k = 0; k < K; ++k) {
+                        // src0 is column major, src1 is row major
+                        int lane_A = i + M * (block + B * (k / K_L));
+                        int lane_B = j + N * (block + B * (k / K_L));
+                        int item = k % K_L;
+                        result[i][j] += src0[item][lane_A] * src1[item][lane_B];
+                    }
+                }
+            }
+
+            for (int i = 0; i < M; ++i) {
+                for (int j = 0; j < N; ++j) {
+                    int item = (i % H) + H * (i/(H*M_I) + G * (block / B_I));
+                    int lane = j + N * ((i / H) % M_I + M_I * (block % B_I));
+
+                    vdst[item][lane] = result[i][j];
+                }
+            }
+
+            for (int i = 0; i < gprs_c_d; ++i) {
+                vdst[i].write();
```

Comment:
This should be pulled out of the `blocks` loop, otherwise the last block will overwrite the previous blocks.